File: scoutScheduler.py
import cv2
import os
import requests
import json
import tbapy
from operator import itemgetter
import sqlite3 as sql
from pathlib import Path
import xlsxwriter
import math
from datetime import datetime

#Config
TBAkey = "yZEr4WuQd0HVlm077zUI5OWPfYsVfyMkLtldwcMYL6SkkQag29zhsrWsoOZcpbSj"
scoutRecordsDatabase = "testDatabase.db"
saveUpdatedRecords = False
outputFile = "schedule.xlsx"


if saveUpdatedRecords:
    print("Will save updated scout records.")
else:
    print("Will NOT save updated scout records.")
print()

#Check if database exists
tempPath = Path(scoutRecordsDatabase)
createNewDatabase = not tempPath.is_file()

#Connect to database
def connectDB():
    global conn
    global cur
    conn = sql.connect(scoutRecordsDatabase)
    cur = conn.cursor()

scoutRecords = []
scoutRecords_clean = []
def createScoutRecords():
    for i in range(0, len(scoutlist)):
        scoutRecords.append({})
        scoutRecords_clean.append({})

if createNewDatabase: #Create new database
    print("No database found at '" + scoutRecordsDatabase + "'")
    rawScouts = input("Enter a comma-separated list of scouts: ")
    scoutlist = [x.strip() for x in rawScouts.split(',')]
    createScoutRecords()
    connectDB()
    
    cur.execute("""CREATE TABLE matchRecords (
        scout TEXT,
        team TEXT,
        count INTEGAR
        ); """)
    cur.execute("""CREATE TABLE preferences (
        team INTEGAR,
        scout TEXT
        ); """)
    cur.execute("""CREATE TABLE scouts (
        name TEXT,
        enabled INTEGAR
        ); """)
    for i in range(0, len(scoutlist)):
        cur.execute("INSERT INTO scouts(name,enabled) VALUES (?,1)", (scoutlist[i],))
    conn.commit()

    print("Created database. Stop program now to add team preferences.")
    print()

else: #Read records from database
    connectDB()
    #Get scouts
    cur.execute("SELECT * FROM scouts")
    dbScouts = cur.fetchall()
    scoutlist = []
    for row in dbScouts:
        if row[1] == 1:
            scoutlist.append(row[0])

    #Get records
    cur.execute("SELECT * FROM matchRecords")
    dbMatchRecords = cur.fetchall()
    createScoutRecords()
    for row in dbMatchRecords:
        if row[0] in scoutlist:
            scoutRecords[scoutlist.index(row[0])][row[1]] = row[2]
            scoutRecords_clean[scoutlist.index(row[0])][row[1]] = row[2]

#Finish creation of scout records array
for i in range(0, len(scoutRecords)):
    scoutRecords[i]['total'] = 0
    scoutRecords[i]['id'] = i

#Check for 6 scouts
if len(scoutlist) < 6:
    print("Found " + str(len(scoutlist)) + " scouts. Must have 6 scouts.")
    quit()

#Initialize TBA link
tba = tbapy.TBA(TBAkey)

#Get event code
year = input("Enter year: ")
teamEventsRaw = tba.team_events('frc6328', year)
teamEventsSorted = sorted(teamEventsRaw, key=itemgetter('start_date'))
print()
print("Events:")
i = 0
for i in range(0, len(teamEventsSorted)):
    print(str(i + 1) + ') ' + teamEventsSorted[i].name)
print()
eventNumberRaw = input("Event #: ")
eventNumber = int(eventNumberRaw)-1
event = teamEventsSorted[eventNumber].key
eventFriendlyname = teamEventsSorted[eventNumber].name

#Create match schedule
print("Fetching schedule...")
matchlistRaw = tba.event_matches(event)
if len(matchlistRaw) == 0:
    print("No schedule available.")
    quit()

matchlistUnsorted = {}
matchlistDaysUnsorted = {}
for i in range(0, len(matchlistRaw)):
    if matchlistRaw[i].comp_level == 'qm':
        matchlistUnsorted[matchlistRaw[i].match_number] = [matchlistRaw[i].alliances["blue"]["team_keys"][0], matchlistRaw[i].alliances["blue"]["team_keys"][1], matchlistRaw[i].alliances["blue"]["team_keys"][2], matchlistRaw[i].alliances["red"]["team_keys"][0], matchlistRaw[i].alliances["red"]["team_keys"][1], matchlistRaw[i].alliances["red"]["team_keys"][2]]
        matchlistDaysUnsorted[matchlistRaw[i].match_number] = matchlistRaw[i].time

matchlist = []
for i in sorted(matchlistUnsorted.keys()):
    matchlist.append(matchlistUnsorted[i])

matchlistDays = []
day = 0
lastDate = -1
for i in sorted(matchlistDaysUnsorted.keys()):
    matchDay = int(datetime.utcfromtimestamp(matchlistDaysUnsorted[i]).strftime('%d'))
    if matchDay > lastDate:
        day += 1
        lastDate = matchDay
    matchlistDays.append(day)

print("Assigning scouts...")

#Get team list
teamlist = []
for matchnumber in range(0, len(matchlist)):
    for teamnumber in range(0, 6):
        if matchlist[matchnumber][teamnumber] not in teamlist:
            teamlist.append(matchlist[matchnumber][teamnumber])

#Update scout records array (to add teams)
for i in range(0, len(scoutlist)):
    for teamnumber in range(0, len(teamlist)):
        if teamlist[teamnumber] not in scoutRecords[i].keys():
            scoutRecords[i][teamlist[teamnumber]] = 0
            scoutRecords_clean[i][teamlist[teamnumber]] = 0

#Get preferences
cur.execute("SELECT * FROM preferences")
dbPreferences = cur.fetchall()
for row in dbPreferences:
    if row[1] in scoutlist:
        fullTeam = "frc" + str(row[0])
        scoutRecords[scoutlist.index(row[1])][fullTeam] = 9999
    else:
        print("WARNING: not obeying preference for Team " + str(row[0]) + " - scout '" + row[1] + "' not found or disabled")

#Create priority lists
def priorityList(team):
    sortedScouts = sorted(scoutRecords, key=lambda x: (-x[team], x['total']))
    tempOutput = []
    for i in range(0, len(sortedScouts)):
        tempOutput.append(sortedScouts[i]['id'])
    return(tempOutput)

#Create match schedule
def createSchedule(match):
    scheduled = {}
    
    #Generate priority lists
    priorityLists = {}
    for teamnumber in range(0, 6):
        priorityLists[match[teamnumber]] = priorityList(team=match[teamnumber])
    
    #Function for removing a scout from priority lists (once assigned)
    def removeFromPriority(scout):
        for team, list in priorityLists.items():
            while scout in priorityLists[team]:
                priorityLists[team].remove(scout)

    #Function for one cycle of assignments
    def assignScouts():
        #Generate lists of scout requests
        scoutRequests = []
        for i in range(0, len(scoutlist)):
            scoutRequests.append([])
        for team, list in priorityLists.items():
            if len(priorityLists[team]) > 0:
                scoutRequests[priorityLists[team][0]].append(team)
        
        #Iterate through scout requests (resolving conflicts when neccessary)
        for scoutRequestNumber in range(0, len(scoutRequests)):
            if len(scoutRequests[scoutRequestNumber]) == 1:
                #No conflict (scout requested by one team)
                scheduled[scoutRequests[scoutRequestNumber][0]] = scoutRequestNumber #Add to schedule
                priorityLists[scoutRequests[scoutRequestNumber][0]] = [] #Clear priority list for team
                removeFromPriority(scout=scoutRequestNumber) #Remove scout from priority lists (so cannot be selected for another team)
            elif len(scoutRequests[scoutRequestNumber]) > 1:
                #Conflict found (scout requested by multiple teams)
                #Resolved by comparing potential 'loss of experience' if each team used secondary scout
                comparisonData = []
                for i in range(0, len(scoutRequests[scoutRequestNumber])):
                    comparisonData.append(scoutRecords[priorityLists[scoutRequests[scoutRequestNumber][i]][0]][scoutRequests[scoutRequestNumber][i]] - scoutRecords[priorityLists[scoutRequests[scoutRequestNumber][i]][1]][scoutRequests[scoutRequestNumber][i]]) #Find difference between experience of primary and secondary scout
                maxid = 0
                for i in range(0, len(comparisonData)):
                    if comparisonData[i] > comparisonData[maxid]:
                        maxid = i
                scheduled[scoutRequests[scoutRequestNumber][maxid]] = scoutRequestNumber #Add to schedule
                priorityLists[scoutRequests[scoutRequestNumber][maxid]] = [] #Clear priority list for team
                removeFromPriority(scout=scoutRequestNumber) #Remove scout from priority lists (so cannot be selected for another team)
    
    #Run cycles of assignment until schedule created
    while len(scheduled) < 6:
        assignScouts()
    
    #Update scout records
    for team, scout in scheduled.items():
        scoutRecords[scout][team] += 1
        scoutRecords_clean[scout][team] += 1
        scoutRecords[scout]['total'] += 1

    return scheduled

#Create schedule for each match
schedule = []
for matchnumber in range(0, len(matchlist)):
    tempScheduleRaw = createSchedule(matchlist[matchnumber])
    tempSchedule = {}
    for team, scout in tempScheduleRaw.items():
        tempSchedule[int(team[3:])] = scout
    schedule.append(tempSchedule)

#Write scout records to database
if saveUpdatedRecords:
    for i in range(0, len(scoutlist)):
        cur.execute("DELETE FROM matchRecords WHERE scout=?", (scoutlist[i],))
    for scoutnumber in range(0, len(scoutRecords_clean)):
        for team, count in scoutRecords_clean[scoutnumber].items():
            if count > 0:
                cur.execute("INSERT INTO matchRecords(scout,team,count) VALUES (?,?,?)", (scoutlist[scoutnumber],team,count,))
    conn.commit()
    print("Wrote records to database.")

#Write output (create workbook)
workbook = xlsxwriter.Workbook(outputFile)
eventTitle = workbook.add_format({'bold': True, 'font_size': 18})
scoutTitle = workbook.add_format({'underline': True, 'font_size': 14})
headingLeft = workbook.add_format({'bold': True, 'align': 'left'})
headingRight = workbook.add_format({'bold': True, 'align': 'right'})
highlighted = workbook.add_format({'bg_color': 'yellow'})
underlined = workbook.add_format({'underline': True})
dayTitle = workbook.add_format({'underline': True, 'align': 'right'})

#Write output (primary scouts)
worksheet = workbook.add_worksheet("Primary Scouts")
worksheet.write(0, 0, "Team", headingRight)
worksheet.write(0, 1, "Primary", headingLeft)
worksheet.write(0, 2, "Secondary", headingLeft)
primaryScouts = {}
secondaryScouts = {}
for teamnumber in range(0, len(teamlist)):
    sortedScouts = sorted(scoutRecords, key=itemgetter(teamlist[teamnumber]), reverse=True)
    primaryScouts[int(teamlist[teamnumber][3:])] = scoutlist[sortedScouts[0]['id']]
    if sortedScouts[1][teamlist[teamnumber]] == 0:
        secondaryScouts[int(teamlist[teamnumber][3:])] = ""
    else:
        secondaryScouts[int(teamlist[teamnumber][3:])] = scoutlist[sortedScouts[1]['id']]
row = 0
for team in sorted(primaryScouts.keys()):
    row += 1
    worksheet.write(row, 0, team)
    worksheet.write(row, 1, primaryScouts[team])
    worksheet.write(row, 2, secondaryScouts[team])

#Write output (match schedule)
worksheet = workbook.add_worksheet("Matches")
worksheet.write(0, 0, "Match", headingRight)
column = -1
for alliance in ["B", "R"]:
    for i in range(1, 4):
        column += 2
        worksheet.write(0, column, alliance + str(i), headingRight)
        worksheet.write(0, column + 1, "Scout", headingLeft)
for matchnumber in range(0, len(matchlist)):
    worksheet.write(matchnumber + 1, 0, matchnumber + 1)
    for i in range(0, 6):
        worksheet.write(matchnumber + 1, (i * 2) + 1, int(matchlist[matchnumber][i][3:]), underlined)
        worksheet.write(matchnumber + 1, (i * 2) + 2, scoutlist[schedule[matchnumber][int(matchlist[matchnumber][i][3:])]])

#Write output (match schedule (long))
worksheet = workbook.add_worksheet("Matches_long")
worksheet.write(0, 0, "Team", headingRight)
worksheet.write(0, 1, "Match", headingRight)
worksheet.write(0, 2, "Alliance", headingRight)
teamsOutput = []
for matchnumber in range(0, len(matchlist)):
    for i in range(0, len(matchlist[matchnumber])):
        if i < 3:
            alliance = 1
        else:
            alliance = 0
        teamsOutput.append({"match": matchnumber + 1, "team": matchlist[matchnumber][i][3:], "alliance": alliance})
teamsOutput = sorted(teamsOutput, key=lambda x: (x['team'], x['match']))

for i in range(0, len(teamsOutput)):
    worksheet.write(i + 1, 0, int(teamsOutput[i]["team"]))
    worksheet.write(i + 1, 1, teamsOutput[i]["match"])
    worksheet.write(i + 1, 2, teamsOutput[i]["alliance"])

#Write output (scout schedules)
#Create long formatted (sort schedule)
scoutSchedules = {}
for i in range(0, len(scoutlist)):
    scoutSchedules[scoutlist[i]] = []
for matchnumber in range(0, len(schedule)):
    for team, scout in schedule[matchnumber].items():
        scoutSchedules[scoutlist[scout]].append({"match": matchnumber + 1, "team": team})
for scout, matches in scoutSchedules.items():
    scoutSchedules[scout] = sorted(scoutSchedules[scout], key=lambda x: (x['match'], x['team']))

#Generate schedules
for scout, matches in scoutSchedules.items():
    #Set up worksheet
    worksheet = workbook.add_worksheet("Schedule (" + scout + ")")
    worksheet.write(0, 0, "Scouting Schedule (" + eventFriendlyname + ")", eventTitle)
    worksheet.write(1, 0, scout, scoutTitle)
    worksheet.write(3, 0, "Match", headingRight)
    worksheet.write(3, 1, "Team", headingRight)
    
    #Write schedule
    notes = []
    i = -1
    lastDay = 1
    nextWrite = "match"
    matchnumber = 0
    while True:
        i += 1
        column = int(math.floor(i / 43) * 3)
        row = int((i - ((column / 3) * 43)) + 4)
        if matchlistDays[matches[matchnumber]["match"] - 1] > lastDay: #Need to add day title, skip line
            lastDay += 1
            nextWrite = "day"
        
        elif nextWrite == "day": #Add day title
            if row < 42:
                worksheet.write(row, column, "Day " + str(lastDay), dayTitle)
                nextWrite = "match"
        
        else: #Add match
            worksheet.write(row, column, matches[matchnumber]["match"])
            if primaryScouts[matches[matchnumber]["team"]] != scout:
                if not any(d['team'] == matches[matchnumber]["team"] for d in notes): #Add to notes if not already
                    notes.append({"scout": primaryScouts[matches[matchnumber]["team"]], "team": matches[matchnumber]["team"]})
                worksheet.write(row, column+1, matches[matchnumber]["team"], highlighted)
            else:
                worksheet.write(row, column+1, matches[matchnumber]["team"])
            matchnumber += 1
            if matchnumber >= len(matches):
                break

    #Add notes
    if len(notes) > 0:
        if (column / 3) == 2:
            startingColumn = column
            startingRow = row + 2
        else:
            startingColumn = column + 3
            startingRow = 3
        worksheet.write(startingRow, startingColumn, "Team", headingRight)
        worksheet.write(startingRow, startingColumn + 1, "Scout", headingLeft)
        notes = sorted(notes, key=itemgetter("team"))
        for i in range(0, len(notes)):
            worksheet.write(startingRow + i + 1, startingColumn, notes[i]["team"])
            worksheet.write(startingRow + i + 1, startingColumn + 1, notes[i]["scout"])

#Save workbook
workbook.close()
print("Wrote to output file '" + outputFile + "'")

File: scoutScheduler.swift
#!/usr/bin/swift

import Foundation
let key = "yZEr4WuQd0HVlm077zUI5OWPfYsVfyMkLtldwcMYL6SkkQag29zhsrWsoOZcpbSj"
let preferred6328scout: Int? = 0
let outputFile = "/Users/jonah/Desktop/schedule.csv"

func writeString(_ toWrite: String, to destination: String) {
    let url = URL(fileURLWithPath: destination)
    try! toWrite.data(using: .utf8)!.write(to: url, options: .atomic)
}

extension String {
    func toLengthOf(length:Int) -> String {
        if length <= 0 {
            return self
        } else if let to = self.index(self.startIndex, offsetBy: length, limitedBy: self.endIndex) {
            return self.substring(from: to)
            
        } else {
            return ""
        }
    }
}

func runShell(_ args: [String]) -> Data? {
    // Create a Task instance
    let task = Process()
    
    // Set the task parameters
    task.launchPath = "/usr/bin/env"
    task.arguments = args
    
    // Create a Pipe and make the task
    // put all the output there
    let pipe = Pipe()
    task.standardOutput = pipe
    
    // Launch the task
    task.launch()
    
    // Get the data
    let data = pipe.fileHandleForReading.readDataToEndOfFile()
    
    return(data)
}

func TBArequest(_ url: String) -> Data {
    return runShell(["curl", "-s", "-X", "GET", "https://www.thebluealliance.com/api/v3/" + url, "-H", "accept: application/json", "-H", "X-TBA-Auth-Key: " + key])!
}

//Determine event code
print("Enter year:")
let eventlistData = TBArequest("team/frc6328/events/" + readLine()! + "/simple")
let eventlistDataString = String(data: eventlistData, encoding: String.Encoding.utf8) as String!
if eventlistDataString!.range(of: "Invalid year") != nil {
    print()
    print("Invalid year")
    exit(0)
}
let eventlistRaw = try JSONSerialization.jsonObject(with: eventlistData) as! [[String: Any]]
let eventlistSorted = eventlistRaw.sorted(by: { $1["start_date"]! as! String > $0["start_date"]! as! String })

print()
if eventlistSorted.count == 0 {
    print("No events found.")
    exit(0)
}

var id = 0
print("Events:")
for event in eventlistSorted {
    id += 1
    print(String(id) + ") " + String(describing: event["name"]!))
}
print()
print("Which event?")
let eventId = Int(readLine()!)! - 1
let event = String(describing: eventlistSorted[eventId]["key"]!)
let eventName = String(describing: eventlistSorted[eventId]["name"]!)

//Get settings
print()
print("How many scouts?")
let scoutCount = Int(readLine()!)!
if scoutCount < 6 {
    print()
    print("Must have at least 6 scouts.")
    exit(0)
}

//Get match data
print("Creating schedule for " + String(scoutCount) + " scouts at event '" + eventName + "'")
print("Fetching match schedule...")
let data = TBArequest("event/" + event + "/matches/simple")

//Decode JSON
let matchlistRaw = try JSONSerialization.jsonObject(with: data) as! [[String: Any]]
if matchlistRaw.count == 0 {
    print()
    print("No schedule available.")
    exit(0)
}

//Create simple match list
var matchlistQuals: [[String: Any]] = []
for i in matchlistRaw {
    if i["comp_level"]! as! String == "qm" {
        matchlistQuals.append(i)
    }
}
matchlistQuals = matchlistQuals.sorted(by: { $1["match_number"]! as! Int > $0["match_number"]! as! Int })
var matchlist: [[String]] = []
for sourceMatch in matchlistQuals {
    var tempTeams: [String] = []
    let alliances = sourceMatch["alliances"]! as! [String: Any]
    let blueAlliance = alliances["blue"]! as! [String: Any]
    let blueAllianceTeams = blueAlliance["team_keys"]! as! [String]
    for team in blueAllianceTeams {
        tempTeams.append(team)
    }
    let redAlliance = alliances["red"]! as! [String: Any]
    let redAllianceTeams = redAlliance["team_keys"]! as! [String]
    for team in redAllianceTeams {
        tempTeams.append(team)
    }
    matchlist.append(tempTeams)
}

print("Assigning scouts...")

//Create team list
var teamlist: [String] = []
for match in matchlist {
    for team in match {
        if teamlist.contains(team) == false {
            teamlist.append(team)
        }
    }
}

//Create scout record array
var scoutRecords: [[String: Int]] = [[String: Int]](repeating: [:], count: scoutCount)
var currentId = 0
for scout in 0...scoutRecords.count - 1 {
    for team in teamlist {
        scoutRecords[scout][team] = 0
    }
    if preferred6328scout != nil {
        if preferred6328scout == currentId {
            scoutRecords[scout]["frc6328"] = 999
        }
    }
    scoutRecords[scout]["total"] = 0
    scoutRecords[scout]["id"] = currentId
    currentId += 1
}

//Create priority lists
func priorityList(for team: String) -> [Int] {
    let sortedScouts = scoutRecords.sorted(by:{
        if $0[team]! != $1[team]! {
            return $0[team]! > $1[team]!
        } else {
            return $0["total"]! < $1["total"]!
        }
    })
    var tempOutput: [Int] = []
    for i in sortedScouts {
        tempOutput.append(i["id"]!)
    }
    return tempOutput
}

//Create match schedule
func createSchedule(for match: [String]) -> [String: Int] {
    var priorityLists: [String: [Int]] = [:]
    for team in match {
        priorityLists[team] = priorityList(for: team)
    }
    var scheduled: [String: Int?] = [:]
    for team in match {
        scheduled[team] = nil
    }
    func assignScouts() {
        //Generate scout requests
        var scoutRequests: [[String]] = [[String]](repeating: [], count: scoutCount)
        for (team, _) in priorityLists {
            if priorityLists[team]!.count != 0 {
                scoutRequests[priorityLists[team]![0]].append(team)
            }
        }
        
        func removeFromPriority(_ scout: Int) {
            for (team, _) in priorityLists {
                if priorityLists[team]!.count > 0 {
                    let originalCount = priorityLists[team]!.count
                    for f in 1...originalCount {
                        if priorityLists[team]![originalCount - f] == scout {
                            priorityLists[team]!.remove(at: originalCount - f)
                        }
                    }
                }
            }
        }
        for i in 0...scoutRequests.count - 1 {
            if scoutRequests[i].count == 1 {
                scheduled[scoutRequests[i][0]] = i
                priorityLists[scoutRequests[i][0]] = []
                removeFromPriority(i)
            } else if scoutRequests[i].count > 1 {
                var comparisonData: [Int] = []
                for f in scoutRequests[i] {
                    comparisonData.append(scoutRecords[priorityLists[f]![0]][f]! - scoutRecords[priorityLists[f]![1]][f]!)
                }
                var maxid = 0
                for f in 0...comparisonData.count - 1 {
                    if comparisonData[f] > comparisonData[maxid] {
                        maxid = f
                    }
                }
                scheduled[scoutRequests[i][maxid]] = i
                priorityLists[scoutRequests[i][maxid]] = []
                removeFromPriority(i)
            }
        }
    }
    while scheduled.count < 6 {
        assignScouts()
    }
    for (team, scout) in scheduled {
        scoutRecords[scout!][team] = scoutRecords[scout!][team]! + 1
        scoutRecords[scout!]["total"] = scoutRecords[scout!]["total"]! + 1
    }
    var finalSchedule: [String: Int] = [:]
    for (team, scout) in scheduled {
        finalSchedule[team] = scout!
    }
    return finalSchedule
}

var schedule: [[String: Int]] = []
for match in matchlist {
    schedule.append(createSchedule(for: match))
}

//Write to output file
print("Saving to file '\(outputFile)'")
var tempOutput = "Match,Team1,Team2,Team3,Team4,Team5,Team6,Team1scout,Team2scout,Team3scout,Team4scout,Team5scout,Team6scout"
var matchNumber = 0
for match in schedule {
    matchNumber += 1
    var teams: [String] = []
    var scouts: [String] = []
    for (team, scout) in match {
        teams.append(team.toLengthOf(length: 3))
        scouts.append(String(scout + 1))
    }
    var fullList = teams + scouts
    var tempLine = "\n\(matchNumber),"
    for i in 0...fullList.count - 1 {
        tempLine = tempLine + fullList[i]
        if i != fullList.count - 1 {
            tempLine = tempLine + ","
        }
    }
    tempOutput = tempOutput + tempLine
}
writeString(tempOutput, to: outputFile)
print("Saved schedule.")

var matchnumber = 0
for match in schedule {
    matchnumber += 1
    for (team, scout) in match {
        if scout == 0 {
            print("M\(matchnumber): \(team)")
        }
    }
}
